Site/dataobjects/SiteAudioMedia.php

```diff
@@ -279,6 +233,75 @@ protected function processEncoding($file_path, SiteMediaEncoding $encoding)
 	}
 
 	// }}}
+	// {{{ protected function parseMp3Duration()
+
+	protected function parseMp3Duration($file_path, $offset = 0)
+	{
+		$bin = trim(`which ffprobe`);
+
+		$duration = null;
+
+		$command = sprintf(
+			'%s '.
+				'-print_format json '.
+				'-read_intervals %s%% '.
+				'-select_streams a '.
+				'-show_entries packet=pts_time '.
+				'-v quiet '.
+				'%s ',
+			$bin,
+			escapeshellarg($offset),
+			escapeshellarg($file_path)
+		);
+
+		$returned_value = 0;
+		$command_output = '';
+		exec($command, $command_output, $returned_value);
+
+		// If ffprobe has worked, get the time from it's output,
+		// otherwise throw an exception.
+		if ($returned_value === 0) {
+			$result = implode('', $command_output);
+			$result = json_decode($result, true);
+
+			if ($result !== null &&
+				is_array($result['packets']) &&
+				count($result['packets']) > 0) {
+				$packet = end($result['packets']);
+				$duration = (integer)round($packet['pts_time']);
+			}
+
+			if ($duration === null && $offset > 0) {
+				// Depending on the encoding, some MP3s will return no packets
+				// when read_intervals goes past the end of the file. If
+				// no packets are returned, run again and read all packets.
+				// It's slower, but it works.
+				$duration = $this->parseMp3Duration($file_path);
```

Comment:
This could potentially cause a recursive loop.

Reply:
the `&& $offset > 0` is meant to prevent that